src/interpreter/kernel/exec/index.js

```diff
@@ -0,0 +1,215 @@
+
+  function getLocalByIndex(index: number) {
+    const local = frame.locals[index];
+
+    if (typeof local === 'undefined') {
+      throw new RuntimeError('Assertion error: no local value at index ' + index);
+    }
+
+    frame.values.push(local);
+  }
+
+  function setLocalByIndex(index: number, value: StackLocal) {
+    assert(typeof index === 'number');
+
+    frame.locals[index] = value;
+  }
+
+  function pushResult(res: StackLocal) {
+    frame.values.push(res);
+  }
+
+  function popArrayOfValTypes(types: Array<Valtype>): any {
+    assertNItemsOnStack(frame.values, types.length);
+
+    return types.map((type) => {
+      return pop1(type);
+    });
+  }
+
+  function pop1(type: ?Valtype): any {
+    assertNItemsOnStack(frame.values, 1);
+
+    const v = frame.values.pop();
+
+    if (typeof type === 'string' && v.type !== type) {
+      throw new RuntimeError(
+        'Internal failure: expected value of type ' + type
+        + ' on top of the stack, give type: ' + v.type
+      );
+    }
+
+    return v;
+  }
+
+  function pop2(type1: Valtype, type2: Valtype): [any, any] {
+    assertNItemsOnStack(frame.values, 2);
+
+    const c2 = frame.values.pop();
+    const c1 = frame.values.pop();
+
+    if (c2.type !== type2) {
+      throw new RuntimeError(
+        'Internal failure: expected c2 value of type ' + type2
+        + ' on top of the stack, give type: ' + c2.type
+      );
+    }
+
+    if (c1.type !== type2) {
+      throw new RuntimeError(
+        'Internal failure: expected c1 value of type ' + type2
+        + ' on top of the stack, give type: ' + c1.type
+      );
+    }
+
+    return [c1, c2];
+  }
+
+  while (pc < frame.code.length) {
+    const instruction = frame.code[pc];
+
+    switch (instruction.type) {
+
+    /**
+     * Function declaration
+     *
+     * FIXME(sven): seems unspecified in the spec but it's required for the `call`
+     * instruction.
+     */
+    case 'Func': {
+      const func = instruction;
+
+      /**
+       * Register the function into the stack frame labels
+       */
+      if (typeof func.id === 'string') {
+        frame.labels[func.id] = func;
+      }
+
+      break;
+    }
+
+    }
+
+    const frameutils = {
+      assertNItemsOnStack,
+      pop1,
+      pop2,
+      isTrapped,
+      assert,
+      depth,
+      castIntoStackLocalOfType,
+      popArrayOfValTypes,
+      pushResult,
+      setLocalByIndex,
+      isZero,
+      getLocalByIndex,
+
+      createChildStackFrame,
+      executeStackFrame,
+    };
+
+    function createInstructionEvaluator(visitor: Object) {
+
+      return function evaluate(instruction: Instruction) {
+        let res;
+
+        const fn = visitor[instruction.id];
+
+        if (typeof fn !== 'undefined') {
+          res = fn.bind(visitor)(instruction, frame, frameutils);
+        }
+
+        if (isTrapped(res)) {
+          return res;
+        }
+
+      };
+    }
+
+    const evaluate = createInstructionEvaluator(Object.assign({},
```

Comment:
This is currently not efficient but i'm using the frameutils object in closure. I will change it in this PR.